extension/spark-doris-connector/src/main/java/org/apache/doris/spark/serialization/RowBatch.java

```diff
@@ -0,0 +1,273 @@
+            this.cols = new ArrayList<>(colCount);
+        }
+
+        List<Object> getCols() {
+            return cols;
+        }
+
+        public void put(Object o) {
+            cols.add(o);
+        }
+    }
+
+    private int offsetInOneBatch = 0;
+    private int rowCountInOneBatch = 0;
+    private int readRowCount = 0;
+    private final ArrowStreamReader arrowStreamReader;
+    private final VectorSchemaRoot root;
+    private List<FieldVector> fieldVectors;
+    private RootAllocator rootAllocator;
+    private final Schema schema;
+
+    public RowBatch(TScanBatchResult nextResult, Schema schema) throws DorisException {
+        logger.trace("Generate Row Batch");
+        this.schema = schema;
+        this.rootAllocator = new RootAllocator(Integer.MAX_VALUE);
+        this.arrowStreamReader = new ArrowStreamReader(
+                new ByteArrayInputStream(nextResult.getRows()),
+                rootAllocator
+                );
+        try {
+            this.root = arrowStreamReader.getVectorSchemaRoot();
+        } catch (Exception e) {
+            logger.error("Read Doris Data failed because: ", e);
+            close();
+            throw new DorisException(e.getMessage());
+        }
+    }
+
+    public boolean hasNext() throws DorisException {
+        if (offsetInOneBatch < rowCountInOneBatch) {
+            return true;
+        }
+        try {
+            try {
+                while (arrowStreamReader.loadNextBatch()) {
+                    fieldVectors = root.getFieldVectors();
+                    readRowCount += root.getRowCount();
+                    if (fieldVectors.size() != schema.size()) {
+                        logger.error("Schema size '{}' is not equal to arrow field size '{}'.",
+                                fieldVectors.size(), schema.size());
+                        throw new DorisException("Load Doris data failed, schema size of fetch data is wrong.");
+                    }
+                    if (fieldVectors.size() == 0 || root.getRowCount() == 0) {
+                        logger.debug("One batch in arrow has no data.");
+                        continue;
+                    }
+                    offsetInOneBatch = 0;
+                    rowCountInOneBatch = root.getRowCount();
+                    return true;
+                }
+            } catch (IOException e) {
+                logger.error("Load arrow next batch failed.", e);
+                throw new DorisException("Cannot load arrow next batch fetching from Doris.");
+            }
+        } catch (Exception e) {
+            close();
+            throw e;
+        }
+        return false;
+    }
+
+    public List<Object> next() throws DorisException {
+        try {
+            if (!hasNext()) {
+                String errMsg = "Get row offset:" + offsetInOneBatch + " larger than row size: " + rowCountInOneBatch;
+                logger.error(errMsg);
+                throw new IndexOutOfBoundsException(errMsg);
```

Comment:
Can you explain this logic?

Reply:
NoSuchElementException should better than IndexOutOfBoundsException.
HaxNext return false means no more data in RowBatch.
RowBatch acts like an iterator. So caller should call hasNext first. if hasNext return true, call next to get data from RowBatch.